Integrate REST api during saving

gkaragia opened the pull request:
### Changes proposed in this Pull Request

* The REST integration is implemented when updating the lesson's quiz.
* The settings of the quiz, although they are updated correctly, they are overwritten by the form that already exists in the lesson. The form will be hidden when [this PR](https://github.com/Automattic/sensei/pull/4009) is merged so after this point the quiz settings should work correctly. 

### Testing instructions

* Try creating and updating all different types of questions and observe that the questions are stored correctly.
* Observe that the questions are displayed correctly on the frontend.
* The quiz settings should be tested after one of the following PRs are merged and this branch is rebased: https://github.com/Automattic/sensei/pull/3972 https://github.com/Automattic/sensei/pull/4009

### New/Updated Hooks

* `sensei_rest_api_question_saved`: This action is triggered when a question is created or updated by the lesson quiz REST endpoint.
* `sensei_quiz_mapped_api_attributes`: Filters the attributes that are retrieved from the REST API in the frontend.
* `sensei_quiz_mapped_block_attributes`: Filters the attributes that will be sent to the REST API when updating a block.

jom commented on includes/rest-api/class-sensei-rest-api-lesson-quiz-controller.php:
```diff
@@ -284,7 +284,7 @@ private function get_question_meta( array $question ) : array {
 				}
 
-				if ( isset( $question['answer_feedback'] ) ) {
+				if ( isset( $question['answer_feedback'] ) || array_key_exists( 'answer_feedback', $question ) ) {
```
I think for all these the `isset()` check is now redundant.

gkaragia replied:
Yeah it's a minor performance optimization for the case where the parameter exists but it is null. But it probably makes sense to just keep `array_key_exists`.

gkaragia commented:
I added a couple of filters and one action to enable users defining a custom question type.

jom commented on assets/blocks/quiz/question-block/question-block-attributes.js:
```diff
@@ -34,29 +48,149 @@ export function createQuestionBlockAttributes( question ) {
+			break;
+		default:
+			blockAttributes = {
+				answer: normalizeAttributes( apiAttributes, camelCase ),
```
Why are we putting these in `answer` by default?

gkaragia replied:
So currently, a custom component for a custom question type would require an `edit` component and optionally a selection of component settings (check `answer-blocks/index.js`). The way this works in `question-edit` is that we pass everything under `answer` in the edit component and everything under `options` in the settings components.

If with the defaults we want to achieve a solution that might work, we need to pass it under `answer` otherwise no attributes will be passed to the edit component.

gkaragia replied:
I have tried to document the above a bit in the `sense_quiz_question_types` filter.

jom replied:
Ahh cool. Thanks for the explanation.

jom commented on assets/blocks/quiz/question-block/question-block-attributes.js:
```diff
@@ -34,29 +48,149 @@ export function createQuestionBlockAttributes( question ) {
+	 * Filters the attributes that will be supplied to the question block after they are received by REST API.
+	 *
+	 * @param {string} type            The question type.
+	 * @param {Object} blockAttributes The block attributes.
```
Think these might be flipped

gkaragia replied:
Good catch! Fixed in 13bcec6f37303556763cb53961050a384b65666c

jom commented on assets/blocks/quiz/question-block/question-block-attributes.js:
```diff
@@ -34,29 +48,149 @@ export function createQuestionBlockAttributes( question ) {
+				snakeCase
+			);
+			apiAttributes = {
+				...normalizedAttributes?.answer,
```
Same here as above, but with `options` as well.

jom commented on assets/blocks/quiz/quiz-store.js:
```diff
@@ -129,4 +146,20 @@ registerStructureStore( {
+		return {
+			...structure,
+			questions: structure.questions.map( ( question ) =>
+				omit( question, [ 'categories', 'shared', 'student_help' ] )
```
Good to know this array is here. I wonder about putting somewhere more obvious, like possibly a constant of read-only attributes? Not sure. I remember there was a reason why `student_help` is here, but I don't remember what it was. We'll also add the `editable` flag once we add it the the REST API.

gkaragia replied:
Good idea! Moved them to a constant in 13bcec6f37303556763cb53961050a384b65666c. 

Now that I think about it, it might be more clear if we expose read only attributes as an argument to `registerStructureStore` instead of exposing `setServerStructure`. But maybe exposing the method is more expandable?

`student_help` is an attribute of file upload which we intend to integrate in the question description. Maybe a better handling of it can be implemented when we work on the migration.

jom reviewed (commented):
Few minor things, and none should affect testing so thought I'd post while testing.

jom reviewed (commented):
A few pending issues that I've found so far:
1) If I create a multiple choice question but don't set up any answers, on save the `answers` child block goes away and I can't get it back.
2) On save, all empty questions are removed including auto-empty block and because of that there isn't a way for me to add a new question (even refreshing strangely doesn't fix the auto-block). For mid-quiz questions, I'm not sure what the desired behavior is. Should we keep them in post-content or should we delete on save?
3) Create a lesson, add new questions, save, then change one of the new question titles, save again, refresh -> I get a warning about changes not being saved. (cc: @yscik in case this is just the store; also a store bug: notice shows up on top of the lesson page when creating a new lesson )

gkaragia commented:
Hey @jom, I had a look today in the issues that we investigated on Friday and did the following fixes:
- Modified the structure update request to happen only after post saving. This should resolve any race conditions which ended up the quiz being saved with an auto-draft status.
- Added an extra check in PHP to not save the quiz with auto-draft status as it still possible to happen by using the endpoint in other places.
- Updated the quiz editor to show an 'Add question' block when the block is selected even after updates. Depending on which block is selected during update, sometimes the editor will lose focus which ends up removing the block but this is fine as long as the 'Add question' block is readded when the quiz block is selected again.
- I added an extra fix where an error is displayed when the user tries to save a question with an empty title. Previously the question would simply dissappear which is problematic.

There is a pending fix of the infinite loop for course structure which I intend to fix on another PR.

jom reviewed (approved):
These fixes look great and work well. 🥇 🥇 🥇
